### PR TITLE
fix: discern usage of custom errors inside require for custom-errors rule

### LINKED ISSUE #148: Discern usage of custom errors inside `require` for `custom-errors` rule
Hello! I have a question about this rule: https://github.com/solhint-community/solhint-community/blob/master/docs/rules/best-practises/custom-errors.md

I understand the encouragement to use the custom error functions to throw custom-defined errors, but why is this code a bad practice? 
```solidity
function someFunction(address _someAddress) external {
  require(_someAddress != 0, ZeroAddressError()); // ZeroAddressError - is my custom defined error.
}
```
Note: the custom errors inside the required functions are supported by Solidity with the `--via-ir` argument: https://docs.soliditylang.org/en/latest/ir-breaking-changes.html.

Thank you in advance!
P.S. If this is the wrong place to ask such questions, please let me know :) And don't forget to direct me to the right place 😅

### PR DESCRIPTION
**Pull request by jahabeebs**
closes #148 

- Checks inside require statements to ensure more fine-grained linting now that custom errors inside require are supported

### CODE REVIEW

File: test/rules/best-practises/custom-errors.js [hunk 1]
```diff
@@ -186,4 +186,53 @@ describe('Linter - custom-errors', () => {
...
+    assertNoErrors(report)
+  })
+
+  it('should raise error for require with a non-function-call second argument', () => {
+    // second arg is a numeric literal instead of a function call
+    const code = `
+      pragma solidity 0.8.5;
+      contract A {
+        function test() external {
+          require(msg.sender != address(0), 123);
```
Comment (juanpcapurro) [inline test/rules/best-practises/custom-errors.js:212]: is this allowed by the compiler?

File: lib/rules/best-practises/custom-errors.js [hunk 2]
```diff
@@ -72,16 +72,39 @@ class CustomErrorsChecker extends BaseChecker {
...
+    else if (node.expression.name === 'require') {
+      // If no second argument or second argument is a string, flag it
+      if (node.arguments.length < 2) {
+        // No second argument
+        errorStr = 'require'
+      } else {
+        const secondArg = node.arguments[1]
+        if (secondArg.type === 'StringLiteral') {
+          // e.g. require(cond, "Error message");
+          errorStr = 'require'
```
Comment (juanpcapurro) [inline lib/rules/best-practises/custom-errors.js:96]: why two `if`s if the body for both is the same?

**Review by juanpcapurro — CHANGES_REQUESTED** [review]
I'd feel more confortable merging this if it included more testcases where the `require`s second argument is a more complex expression, not necessarily because we will handle them all in the rule (it's not possible to cover all cases with static analysis), but as a first step towards exploring the different posibilities and figuring out what we want the rule to do in each case.

**Review by juanpcapurro — CHANGES_REQUESTED** [review]
while it is likely to see little real-world execution, we should not surprise users with confusing & undocumented behaviour such as reporting on a line that has a non-function-call expression _which might actually resolve to a custom error_  as a require's second argument with an error message saying 'use custom errors'

To mitigate that, I believe we should
- [x] use a few different error messages stating _why_ we are reporting on a line. We should not disallow the use of require in our wording, since it's now possible to use custom errors in them.
- [x] IIRC the way we interpret doc's metadata allows for an 'extra comments' or similar field, where we could document the limitations of this approach around accepting any function call as a custom error and the reason why we disallow complex expressions as a require's second parameter

**Comment by juanpcapurro** [comment]
@0xVolosnikov we are in the process of implementing this, your input on the rule's expected behaviour would be hugely appreciated :sparkles:

File: test/rules/best-practises/custom-errors.js [hunk 3]
```diff
@@ -109,81 +116,399 @@ describe('Linter - custom-errors', () => {
...
+        }
+      }
+    `
+    const report = linter.processStr(code, {
+      rules: { 'custom-errors': 'error' },
+    })
+    assertNoErrors(report)
+  })
+
+  it('should handle errors defined in interfaces', () => {
```
Comment (juanpcapurro) [inline test/rules/best-practises/custom-errors.js:452]: This test and the two above are testing the same thing, as far as the rule is concerned, since the AST (only data input for the rule) is equivalent around the require. The `FunctionCall` node does not have any information on its expression that is not present at the grammar level, such as if the name came from a previous in-file definition, is imported, is not actually defined anywhere, etc.

I don't personally mind leaving them in, but as a learning excercise you could inspect the AST for yourself, verify this and delete the tests you consider to be duplicates. You can also try replacing the names defined in the code snippet for unrelated, undefined ones and you'll see it makes no difference to the rule 🙃 

...you can probably get the rule to eat up fairly cursed productions, such as `require(false, require(true, require(false,foo())))` which are perfectly valid gramatically, but make zero semantic sense and are rejected by the compiler at a later stage. This is to illustrate only, you can have fun with that, but I don't expect we test it.

**Review by juanpcapurro — COMMENTED** [review]
This is a very clear improvement from the previous iteration. It even uncovered existing, unrelated edge cases :sweat_smile:

While this is at a mergeable quality, I believe we can go the extra mile and document the fact that we ignore assembly reverts, and also (kinda optionally) make sure we are not doing anything disruptive in the case where a user incorrectly uses revert: `revert(CustomError())`.

I'm concerned we are reporting an error on the above scenario, which would be annoying to users since they might try to correct _the wrong thing_ when their editor reports linting errors alongside ones reported by the compiler. See Uniform Theory of Linting Issues below:

|                | compiles | does not compile |
| ---            | ---      | ---              |
| false positive | bad      | annoying         |
| false negative | BAD      | okay             |

- false positive which compiles: a rule reporting when it shouldn't on code which compiles is not catastrophic, but they will persistently show users how poorly implemented your linter is, as they will most of the time be able to figure out the rule tries to prevent something else and disregard the warning. example: #159
- false negatives on code which compiles are BAD, very BAD, since they can potentially allow machine-detectable issues to sneak through.
- false negatives on code which does *not* compile are okay. We are not preventing the user from doing something which might be wrong, but they will figure it out soon enough when solc burps out its own error message which they carefully crafted in very smart C++ so we don't have to. We might ship a shoddy product, but no-one will notice in this case :smiling_imp:   eg: missing `_` in modifiers
- false positives on code which does *not* compile are kinda annoying, since in this case you'd get an error saying 'hey you have to use custom errors in X way instead of Y' when the code is actually not compiling because reverting with custom errors has to be done using the revert statement syntax instead of the function-call-but-its-a-builtin syntax. And `solc` might give another error whose wording is contradictory to what the rule is saying.

tl;dr document we ignore assembly reverts and make sure we are not inadvertedly sending junior devs off a cliff when they `revert(CustomError())`

**Review by juanpcapurro — APPROVED** [review]
great work, thanks for this
